File: backend/src/source-ir.ts
import {createHash} from 'node:crypto';
import {TextDecoder} from 'node:util';
import {Page} from './entities';
import {AiError} from './core';

type Span={pageIndex:number;startByte:number;endByte:number};
type Mapping={normalizedStartByte:number;normalizedEndByte:number;source:Span;operation:'IDENTITY'|'WHITESPACE'};
export type SourceIR={schemaVersion:'source-ir-v1';stage:'SOURCE_ONLY';documentId:string;parserVersion:string;normalizerVersion:'whitespace-map-v1';sourceHash:string;pages:Page[];blocks:{blockId:string;kind:'paragraph'|'separator';source:Span;normalizedText:string;mapping:Mapping[]}[];irHash:string};
const canonical=(v:any):string=>JSON.stringify(v&&typeof v==='object'?Array.isArray(v)?v.map(x=>JSON.parse(canonical(x))):Object.fromEntries(Object.keys(v).sort().map(k=>[k,JSON.parse(canonical(v[k]))])):v);
export const irHash=(value:unknown)=>createHash('sha256').update(canonical(value)).digest('hex');
const decoder=new TextDecoder('utf-8',{fatal:true});
export function validateSourceIR(input:unknown,documentId:string,pages:Page[]):asserts input is SourceIR {
  const fail=()=>{throw new AiError('SOURCE_IR_INVALID');};
  try {
    const keys=(v:unknown,expected:string[])=>{if(!v||typeof v!=='object'||Array.isArray(v)||Object.keys(v).sort().join(',')!==expected.sort().join(','))fail();};
    const ir=input as SourceIR;
    keys(ir,['schemaVersion','stage','documentId','parserVersion','normalizerVersion','sourceHash','pages','blocks','irHash']);
    if(!ir||ir.schemaVersion!=='source-ir-v1'||ir.stage!=='SOURCE_ONLY'||ir.normalizerVersion!=='whitespace-map-v1'||typeof ir.parserVersion!=='string'||ir.documentId!==documentId||!Array.isArray(ir.pages)||canonical(ir.pages)!==canonical(pages)||ir.sourceHash!==irHash(pages))fail();
    const {irHash:hash,...body}=ir;
    if(hash!==irHash(body)||!Array.isArray(ir.blocks))fail();
    for(const page of ir.pages)keys(page,['pageNumber','text']);
    const source=pages.map(p=>Buffer.from(p.text));
    const ends=pages.map(()=>0),ids=new Set<string>();
    const resolve=(span:Span)=>{
      keys(span,['pageIndex','startByte','endByte']);
      if(!span||![span.pageIndex,span.startByte,span.endByte].every(Number.isSafeInteger)||span.pageIndex<0||span.pageIndex>=source.length||span.startByte<0||span.endByte<=span.startByte||span.endByte>source[span.pageIndex].length)fail();
      return decoder.decode(source[span.pageIndex].subarray(span.startByte,span.endByte));
    };
    let pageOrder=0;
    for(const block of ir.blocks){
      keys(block,['blockId','kind','source','normalizedText','mapping']);
      if(typeof block.blockId!=='string'||ids.has(block.blockId)||!['paragraph','separator'].includes(block.kind)||typeof block.normalizedText!=='string'||!Array.isArray(block.mapping)||!block.mapping.length)fail();
      ids.add(block.blockId);resolve(block.source);
      const page=block.source.pageIndex;
      if(page<pageOrder||block.source.startByte!==ends[page])fail();
      pageOrder=page;
      let offset=block.source.startByte,normalizedOffset=0,text='';
      for(const map of block.mapping){
        keys(map,['normalizedStartByte','normalizedEndByte','source','operation']);
        const raw=resolve(map.source);
        if(map.source.pageIndex!==page||map.source.startByte!==offset||map.source.endByte>block.source.endByte||map.normalizedStartByte!==normalizedOffset)fail();
        if(!['IDENTITY','WHITESPACE'].includes(map.operation))fail();
        if(map.operation==='WHITESPACE'&&!/^[\s\u0085\u001c-\u001f]+$/u.test(raw))fail();
        const rendered=map.operation==='WHITESPACE'?' ':raw;
        text+=rendered;normalizedOffset+=Buffer.byteLength(rendered);offset=map.source.endByte;
        if(map.normalizedEndByte!==normalizedOffset)fail();
      }
      if(offset!==block.source.endByte||text!==block.normalizedText)fail();
      ends[page]=offset;
    }
    if(source.some((buffer,i)=>ends[i]!==buffer.length))fail();
  } catch {fail();}
}

File: ai-service/medical_ai/source_ir.py
"""Lossless source IR foundation. Normalization never changes stored source pages."""
import hashlib
import json
import re
from typing import Literal

from pydantic import Field, model_validator

from .schemas import Page, StrictModel

IR_VERSION = 'source-ir-v1'
NORMALIZER_VERSION = 'whitespace-map-v1'


def canonical_hash(value):
    return hashlib.sha256(json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(',', ':')).encode()).hexdigest()


class Span(StrictModel):
    pageIndex: int = Field(ge=0)
    startByte: int = Field(ge=0)
    endByte: int = Field(ge=0)


class Mapping(StrictModel):
    normalizedStartByte: int = Field(ge=0)
    normalizedEndByte: int = Field(ge=0)
    source: Span
    operation: Literal['IDENTITY', 'WHITESPACE']


class Block(StrictModel):
    blockId: str
    kind: Literal['paragraph', 'separator'] = 'paragraph'
    source: Span
    normalizedText: str
    mapping: list[Mapping]


class SourceIR(StrictModel):
    schemaVersion: Literal['source-ir-v1'] = IR_VERSION
    stage: Literal['SOURCE_ONLY'] = 'SOURCE_ONLY'
    documentId: str
    parserVersion: str
    normalizerVersion: Literal['whitespace-map-v1'] = NORMALIZER_VERSION
    sourceHash: str
    pages: list[Page]
    blocks: list[Block]
    irHash: str

    @model_validator(mode='after')
    def verify(self):
        expected = build_source_ir(self.documentId, self.pages, self.parserVersion)
        if self.model_dump() != expected:
            raise ValueError('IR integrity mismatch')
        return self


def resolve_span(pages: list[Page], span: Span) -> str:
    if span.pageIndex >= len(pages) or span.endByte < span.startByte:
        raise ValueError('Invalid source span')
    raw = pages[span.pageIndex].text.encode()
    if span.endByte > len(raw):
        raise ValueError('Source span out of bounds')
    return raw[span.startByte:span.endByte].decode('utf-8', errors='strict')


def build_source_ir(document_id: str, pages: list[Page], parser_version: str) -> dict:
    """Deterministic paragraphs and complete byte mapping, including all whitespace."""
    blocks = []
    for page_index, page in enumerate(pages):
        # Preserve separator blocks too: no source byte disappears during normalization.
        base = 0
        for number, raw in enumerate(re.split(r'(\n[ \t]*\n)', page.text)):
            if not raw:
                continue
            mappings, normalized = [], []
            source_offset = base
            normalized_offset = 0
            segments = []
            cursor = 0
            for whitespace in re.finditer(r'\s+', raw):
                if whitespace.group() == ' ':
                    continue
                if whitespace.start() > cursor:
                    segments.append((raw[cursor:whitespace.start()], 'IDENTITY'))
                segments.append((whitespace.group(), 'WHITESPACE'))
                cursor = whitespace.end()
            if cursor < len(raw):
                segments.append((raw[cursor:], 'IDENTITY'))
            for token, operation in segments:
                rendered = ' ' if operation == 'WHITESPACE' else token
                source_end = source_offset + len(token.encode())
                normalized_end = normalized_offset + len(rendered.encode())
                mappings.append({'normalizedStartByte': normalized_offset, 'normalizedEndByte': normalized_end,
                    'source': {'pageIndex': page_index, 'startByte': source_offset, 'endByte': source_end},
                    'operation': operation})
                normalized.append(rendered)
                source_offset, normalized_offset = source_end, normalized_end
            blocks.append({'blockId': f'p{page_index}:b{number}', 'kind': 'separator' if raw.isspace() else 'paragraph',
                'source': {'pageIndex': page_index, 'startByte': base, 'endByte': base + len(raw.encode())},
                'normalizedText': ''.join(normalized), 'mapping': mappings})
            base += len(raw.encode())
    raw_pages = [p.model_dump() for p in pages]
    payload = {'schemaVersion': IR_VERSION, 'stage': 'SOURCE_ONLY', 'documentId': document_id,
               'parserVersion': parser_version, 'normalizerVersion': NORMALIZER_VERSION,
               'sourceHash': canonical_hash(raw_pages), 'pages': raw_pages, 'blocks': blocks}
    return {**payload, 'irHash': canonical_hash(payload)}
